### ISSUE TITLE
`no-restricted-imports`: setting to to allow renames

### ISSUE THREAD

**Issue by OliverJAsh** [eslint/eslint#12152]
**What rule do you want to change?**
no-restricted-imports

**Does this change cause the rule to produce more or fewer warnings?**
Fewer, if a setting is enabled

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option

**Please provide some example code that this change will affect:**

```js
{
  rules: {
    'no-restricted-imports': [
      'error',
      {
        paths: [{ name: 'lib', importNames: ['Option'] }],
      },
    ],
  },
}
```

```js
import { Option as LibOption } from 'lib';
```

**What does the rule currently do for this code?**

Errors

**What will the rule do after it's changed?**

Not error, if a setting is enabled, e.g. `allowRenames: true`.

I want to disallow this code:

```js
import { Option } from 'lib';
```

But allow this code:

```js
import { Option as LibOption } from 'lib';
```

That is, I want to ban importing the named import `Option` from `lib`, but I want to allow that named import to be used if it is renamed to something else.

Example of using this setting:

```js
{
  rules: {
    'no-restricted-imports': [
      'error',
      {
        paths: [{ name: 'lib', importNames: ['Option'], allowRenames: true }],
      },
    ],
  },
}
```

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by mdjermanovic** [comment]
Hi @OliverJAsh,

You want to import `Option`, but you don't want to have a variable named `Option` in your code?

This looks to me more like a case for the [`id-blacklist`](https://eslint.org/docs/rules/id-blacklist) rule.

```js
/*eslint id-blacklist: ["error", "Option"] */

import { Option } from 'lib'; // error
import { Option as LibOption } from 'lib'; // (*) shouldn't be error
```

(*) is currently reported as error, I believe it's a bug.

**Comment by OliverJAsh** [comment]
That won't work for us because we do want to allow `Option` identifiers to appear elsewhere—just not when importing from this specific module.

The reason I want this is because we're migrating from one library to another, and we want the old library to be named as such.

This is what we want:

``` ts
import { Option } from 'old-lib'; // error
import { Option as OldLibOption } from 'old-lib'; // error
import { Option } from 'new-lib'; // no error
```

**Comment by mdjermanovic** [comment]
Thanks for the clarification!  

I believe it isn't possible to enforce this with the existing set of rules/options, so it's open for evaluation now.

**Comment by platinumazure** [comment]
Personally, I lean towards this being an edge case, and would suggest creating a custom rule using [eslint-rule-composer](https://github.com/not-an-aardvark/eslint-rule-composer) which could filter out linting reports for your specific case.

That's just my opinion; other team members might feel differently.

**Comment by kaicataldo** [comment]
I agree with @platinumazure. This feels like a muddying of responsibilities of this rule. If you know what identifiers you want to allow, this should also be able to be enforced with [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax).
